src/core/bpf-foreign.c

```diff
@@ -0,0 +1,213 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include "bpf-foreign.h"
+#include "cgroup.h"
+#include "fd-util.h"
+#include "memory-util.h"
+#include "set.h"
+#include "strv.h"
+#include "string-util.h"
+#include "path-util.h"
+
+typedef struct BPFForeignKey BPFForeignKey;
+struct BPFForeignKey {
+        uint32_t prog_id;
+        uint32_t attach_type;
+};
+
+static int bpf_foreign_key_new(uint32_t prog_id,
+                enum bpf_attach_type attach_type,
+                BPFForeignKey **ret) {
+        _cleanup_free_ BPFForeignKey *p = NULL;
+
+        assert(ret);
+
+        p = new(BPFForeignKey, 1);
```

Comment:
there isn't strictly a need to allocate separate heap memory for this here. you could also embed the structure into BPFProgram and then use either the embedded structure or one quickly allocated on the stack for this...

but not sure it's worth the optimization, i.e. what you have now here works fine too i guess, hence leave it as is.